plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php

```diff
@@ -351,27 +393,124 @@ public function cleanup_synchronization_state() {
+	 */
+	private function process_deleted_orders( array $batch, bool $custom_orders_table_is_authoritative ): array {
+		global $wpdb;
+
+		$deleted_from_table_name = $custom_orders_table_is_authoritative ? $this->data_store::get_orders_table_name() : $wpdb->posts;
+		$data_store_for_deletion =
+			$custom_orders_table_is_authoritative ?
+			new \WC_Order_Data_Store_CPT() :
+			wc_get_container()->get( OrdersTableDataStore::class );
+
+		$order_ids_as_sql_list = '(' . implode( ',', $batch ) . ')';
+
+		$deleted_order_ids  = array();
+		$meta_ids_to_delete = array();
+
+		// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+		$deletion_data = $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT id, order_id FROM {$wpdb->prefix}wc_orders_meta WHERE meta_key=%s AND meta_value=%s AND order_id IN $order_ids_as_sql_list ORDER BY order_id DESC",
+				'deleted_from',
+				$deleted_from_table_name
+			),
+			ARRAY_A
+		);
+		// phpcs:enable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+
+		if ( empty( $deletion_data ) ) {
+			return array();
+		}
+
+		foreach ( $deletion_data as $item ) {
+			$meta_id  = $item['id'];
+			$order_id = $item['order_id'];
+
+			if ( isset( $deleted_order_ids[ $order_id ] ) ) {
+				$meta_ids_to_delete[] = $meta_id;
+				continue;
+			}
+
+			if ( ! $data_store_for_deletion->order_exists( $order_id ) ) {
+				$this->error_logger->warning( "Order {$order_id} doesn't exist in the backup table, thus it can't be deleted" );
+				$deleted_order_ids[]  = $order_id;
+				$meta_ids_to_delete[] = $meta_id;
+				continue;
+			}
+
+			try {
+				$order = new \WC_Order();
+
+				// We need to turn parent order verification off to avoid an exception for child orders (e.g. refunds).
+				// The 'read' method in the order class invokes 'set_parent_id', which by default verifies
+				// if the order the parent id refers to actually exists, and throws an exception if not.
+				// To check if the order exists 'wc_get_order' is used, but this function searches the order
+				// using the authoritative data store, and at this point the order doesn't exist anymore in the
+				// authoritative table ('read' below is being invoked in the backup data store).
+				$order->set_verify_parent_id( false );
+
+				$order->set_id( $order_id );
+				$data_store_for_deletion->read( $order );
+
+				$data_store_for_deletion->delete( $order, array( 'force_delete' => true ) );
+			} catch ( \Exception $ex ) {
+				$this->error_logger->error( "Couldn't delete order {$order_id} from the backup table: {$ex->getMessage()}" );
+				continue;
+			}
+
+			$deleted_order_ids[]  = $order_id;
+			$meta_ids_to_delete[] = $meta_id;
+		}
+
+		if ( ! empty( $meta_ids_to_delete ) ) {
+			$order_id_rows_as_sql_list = '(' . implode( ',', $meta_ids_to_delete ) . ')';
+			// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+			$wpdb->query( "DELETE FROM {$wpdb->prefix}wc_orders_meta WHERE id IN {$order_id_rows_as_sql_list}" );
```

Comment:
I know that only meta_key here would be a delete registry, but let's add a meta_key check anyways, incase someone has their own version of delete_registry implemented.

Reply:
On a second thought, this doesn't seem necessary. `$meta_ids_to_delete` is populated from the `$deletion_data` array that is obtained earlier in the same method using a SELECT query that is already filtering by `meta_key`. Or am I missing something?